sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/pubsub/PubsubMessage.java

```diff
@@ -31,51 +31,68 @@
   "nullness" // TODO(https://issues.apache.org/jira/browse/BEAM-10402)
 })
 public class PubsubMessage {
+  @AutoValue
+  abstract static class Impl {
+    @SuppressWarnings("mutable")
+    abstract byte[] getPayload();
 
-  private byte[] message;
-  private @Nullable Map<String, String> attributes;
-  private @Nullable String messageId;
+    abstract @Nullable Map<String, String> getAttributeMap();
+
+    abstract @Nullable String getMessageId();
+
+    static Impl create(
+        byte[] payload, @Nullable Map<String, String> attributes, @Nullable String messageId) {
+      return new AutoValue_PubsubMessage_Impl(payload, attributes, messageId);
+    }
+  }
+
+  private Impl impl;
```

Comment:
I'm curious what this change is for. Is it so you can leverage the AutoValue generated equals and hashCode?

Reply:
Yes, thats exactly what its for. This can't be changed to an AutoValue class without an API break as it has constructors, but we can implement these methods in an automated fashion.